### DISCUSSION TITLE
[Host] Suggest adding a function to obtain the local IP address

### DISCUSSION THREAD

**Discussion by 1101728133** [alec1o/Netly#28] [Q&A (Question and Answer)] [answered]
```
private string GetLocalIpv4()
  {
    string ipv4 = "";
 
      using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, 0))
      {
        socket.Connect("8.8.8.8", 65530);
        ipv4 = (socket.LocalEndPoint as IPEndPoint).Address.ToString();
      }

      return ipv4;
  }
```
Or use other methods，
I don't want to manually input the IP address myself.

**Comment by alec1o** [comment]
Do you wish get internal ip from ``Server`` or ``Client``?

**Reply by 1101728133** [reply]
yes

**Reply by 1101728133** [reply]
Mainly on the server side

**Reply by alec1o** [reply]
Do you try use use ``UdpServer.Host`` ``UdpClient.Host`` ``TcpServer.Host`` ``TcpClient.Host``

https://netly.docs.kezero.com/#/host
```js
var server = new UdpServer(false);

IPAddress address           = server.Host.Address;
IPEndPoint ipEndpoint       = server.Host.IPEndPoint;
EndPoint endpoint           = server.Host.EndPoint;
Int port                    = server.Host.Port;
AddressFamily addressFamily = server.Host.AddressFamily;
```

**Reply by 1101728133** [reply]
![image](https://github.com/alec1o/Netly/assets/55173270/5575c97e-97eb-481f-b361-30a5131fa4f2)
not work.
I hope Netly can provide a built-in method for obtaining local IP addresses

**Reply by 1101728133** [reply]
To avoid having overly cumbersome scripts

**Reply by 1101728133** [reply]
![image](https://github.com/alec1o/Netly/assets/55173270/21150a32-3104-459c-9ee8-b9aed5027bcd)

**Reply by 1101728133** [reply]
There is another issue, if you exit the program, the port will not close and will be occupied. You need to
add ---private void OnApplicationQuit() => server.Close();---
Can Netly automatically handle this issue

**Reply by alec1o** [reply, marked as answer]
lol. The error is because you get Host from not inited server.
- You must connect your server first

<sub>
Well. When a Udp instance is created the Host is null because it has no past connection. So when you try to use udp.Host the value will be null. To solve this, your UDP must have already been connected at least once, in your example it is the same:
</sub>
<br>

```cs
private string GetLocalIpv4()
  {
    string ipv4 = "";
 
      using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, 0))
      {
        socket.Connect("8.8.8.8", 65530); // if you remove this line, will have null ref exception
        ipv4 = (socket.LocalEndPoint as IPEndPoint).Address.ToString();
      }

      return ipv4;
  }
```
<sub>
You needed to connect to get the local IP, the difference is that your solution always needs to connect to get the IP, however netly backs up the host in use or that was in use. <br/>
Don't get it wrong, you can get the Host with the connection disconnected, but your udp must at least have connected once so that netly has some endpoint to return, otherwise netly cannot guess which Host to return once the connection has never been opened with this instance.
</sub>

#### It can resolve your problem.

```cs
server.Open(<Host>);

// or

// print without error, it will print default host instead of get exception
print((server.Host ?? new Host("1.1.1.1", 1).Address.ToString())
```

#### Notice: <sub>The null ref error has fixed: 7db293a now you can use ``server.Host`` by default will return "0.0.0.0:0"</sub>

**Comment by alec1o** [comment]
#### Hello!
<sub>For get internal socket use ``OnModify``. And for get IPAddress, Port, and others. you can also use ``<instance>.Host.Port;``  ``<instance>.Host.IPEndPoint;`` ``...``</sub>

###### Example
```cs
using Netly;
using Netly.Core;
using System.Net.Sockets;

public class Example
{
    private UdpServer server = new UdpServer(false);
    private Socket _socket = null;

    private void Start()
    {
        server.OnOpen(() => print("Server started from"));
        server.OnClose(() => print("Server closed!"));
        server.OnError((e) => print($"Server error: {e}"));

        server.OnModify((socket) => { _socket = socket; });
    }

    private IPEndPoint GetLocalEndPoint()
    {
        return (IPEndPoint)_socket.LocalEndPoint;
    }

    private IPEndPoint GetRemoteEndPoint()
    {
        return (IPEndPoint)_socket.RemoteEndPoint;
    }
}
```

**Reply by alec1o** [reply]
Open new discussion about:
```
There is another issue, if you exit the program, the port will not close and will be occupied. You need to
add ---private void OnApplicationQuit() => server.Close();---
Can Netly automatically handle this issue
```

**Reply by 1101728133** [reply]
I see, it's different from what I thought...QAQ

**Reply by alec1o** [reply]
#### Thank's to open new discusssion. 
##### The answer is released on: [if you exit the program, the port will not close and will be occupied #29](https://github.com/alec1o/Netly/discussions/29)

**Reply by 1101728133** [reply]
You misunderstood, I'm not angry. If it doesn't work, I can change it myself

**Comment by alec1o** [comment]
I don't understand! It's https://github.com/alec1o/Netly/discussions/28#discussioncomment-7869733 working?

**Comment by alec1o** [comment]
### Fixed: [Host is null by default in TCP and UDP #30](https://github.com/alec1o/Netly/issues/30)

#### Problem: 
![image](https://github.com/alec1o/Netly/assets/55173270/5575c97e-97eb-481f-b361-30a5131fa4f2)
not work.
I hope Netly can provide a built-in method for obtaining local IP addresses

**Comment by alec1o** [comment]
https://github.com/alec1o/Netly/discussions/28#discussioncomment-7869733 It's working?

**Reply by 1101728133** [reply]
nice,is working
